python/mxnet/gluon/trainer.py

```diff
@@ -170,11 +178,63 @@ def step(self, batch_size, ignore_stale_grad=False):
         if not self._kv_initialized:
             self._init_kvstore()
 
+        self._allreduce(batch_size)
+        self._update(batch_size, ignore_stale_grad)
+
+    def allreduce(self, batch_size):
+        """For each parameter, reduce the gradients from different contexts. Should be called after
+        `autograd.grad`, outside of `record()` scope, and before `trainer.update`.
+        """
+        if not self._kv_initialized:
+            self._init_kvstore()
+        assert not (self._kvstore and self._update_on_kvstore), \
+                'allreduce() when parameters are updated on kvstore ' \
+                'is not supported. Try setting `update_on_kvstore` ' \
+                'to False when creating trainer.'
+
+        self._allreduce(batch_size)
+
+    def _allreduce(self, batch_size):
+        self._optimizer.rescale_grad = self._scale / batch_size
```

Comment:
It just came to me that there is a bug with Gluon distributed training. when we set rescale_grad after kvstore.set_optimizer, it won't have any effect.

We need an issue to track this